File: TestIOS/SpineDemo/SpineDemo/Spine/BaseTime.cpp
#include "BaseTime.h"

BaseTime::BaseTime()
	: m_timePoint()
	, m_baseTimeUs()
{
	setBaseTimeUs(0);//预先初始化下m_timePoint，只计算两个点的时差，可以直接使用
}

BaseTime::~BaseTime()
{
}

void BaseTime::setBaseTime(int64_t baseTime)
{
	setBaseTimeMs(baseTime * 1000);
}

void BaseTime::setBaseTimeMs(int64_t baseTimeMs)
{
	setBaseTimeUs(baseTimeMs * 1000);
}

void BaseTime::setBaseTimeUs(int64_t baseTimeUs){
	m_baseTimeUs = baseTimeUs;
	m_timePoint = std::chrono::high_resolution_clock::now();//稳定时钟，更改系统时间，不影响此时钟
}

int64_t BaseTime::getBaseTime() const
{
	return getBaseTimeMs() / 1000;
}

int64_t BaseTime::getBaseTimeMs() const
{
	return getBaseTimeUs() / 1000;
}

int64_t BaseTime::getBaseTimeUs()const{
	return m_baseTimeUs;
}

int64_t BaseTime::getCurrentTime()
{
	return getCurrentTimeMs() / 1000;
}

int64_t BaseTime::getCurrentTimeMs()
{
	return getCurrentTimeUs() / 1000;
}

int64_t BaseTime::getCurrentTimeUs(){
	//计算时钟的差
	auto diff = std::chrono::high_resolution_clock::now() - m_timePoint;
	std::chrono::microseconds diffUs = std::chrono::duration_cast<std::chrono::microseconds>(diff);
	return m_baseTimeUs + diffUs.count();
}

File: android/SpineDemo/app/src/main/java/com/ksstory/spinewrap/BaseTime.java
package com.ksstory.spinewrap;

import android.os.SystemClock;

/**
 * 本地无关的时间计算类
 */
public class BaseTime {
    public BaseTime(){
        setBaseTimeUs(0);
    }

    //设置UTC基准时间，单位秒
    public void setBaseTime(long baseTime){
        setBaseTimeMs(baseTime * 1000);
    }

    //设置UTC基准时间，单位毫秒
    public void setBaseTimeMs(long baseTimeMs){
        setBaseTimeUs(baseTimeMs * 1000);
    }

    //设置UTC基准时间，单位微秒
    public void setBaseTimeUs(long baseTimeUs){
        mBaseTimeUs = baseTimeUs;
        mTimePoint = SystemClock.elapsedRealtimeNanos() / 1000;
    }

    //获取UTC基准时间，单位秒
    public long getCurrentTime(){
        return getCurrentTimeMs() / 1000;
    }

    //获取UTC基准时间，单位毫秒
    public long getCurrentTimeMs(){
        return getCurrentTimeUs() / 1000;
    }

    //获取UTC基准时间，单位微秒
    public long getCurrentTimeUs(){
        long diff = SystemClock.elapsedRealtimeNanos() - mTimePoint;
        long diffUs = diff / 1000;
        return mBaseTimeUs + diffUs;
    }

    private long mTimePoint;
    private long mBaseTimeUs;
}
